KEP-4247(blog): QueueingHint (#48541)
kubernetes/website: content/en/blog/_posts/2024-12-12-scheduler-queueinghint/index.md

Reviewer: I made some structural suggestions to improve the flow of information on the page.

---
layout: blog
title: "Kubernetes v1.32: QueueingHint Brings a New Possibility to Optimize Pod Scheduling"
date: 2024-12-12
slug: scheduler-queueinghint
draft: true
Author: >
  [Kensei Nakada](https://github.com/sanposhiho) (Tetrate.io)
---

The Kubernetes [scheduler](/docs/concepts/scheduling-eviction/kube-scheduler/) is the core
component that decides which node any new Pods should run on.
Basically, it schedules Pods **one by one**, 

> and thus the larger your cluster is, the more crucial the throughput of the scheduler is.

Reviewer: nit, really

For the Kubernetes project, the throughput of the scheduler has been an eternal challenge
over the years, SIG Scheduling have been putting effort to improve the scheduling throughput by many enhancements.

In this blog post, I'll introduce a recent major improvement in the scheduler: a new
[scheduling context element](/docs/concepts/scheduling-eviction/scheduling-framework/#extension-points)
named _QueueingHint_.
We'll go through the explanation of the basic background knowledge of the scheduler,
and how QueueingHint improves our scheduling throughput.

## Scheduling queue

The scheduler stores all unscheduled Pods in an internal component that we - SIG Scheduling - 
call the _scheduling queue_.

> The scheduling queue is composed of three data structures: _ActiveQ_, _BackoffQ_ and _Unschedulable Pod Pool_.

Reviewer: no need to say the names, you're saying them in the list

- ActiveQ: It holds newly created Pods or Pods which are ready to be retried for scheduling.
- BackoffQ: It holds Pods which are ready to be retried, but are waiting for a backoff period, which depends on the number of times the scheduled attempted to schedule the Pod.

> - Unschedulable Pod Pool: It holds Pods which should not be scheduled for now, because they have a Scheduling Gate or because the scheduler attempted to schedule them and nothing has changed in the cluster that could make the Pod schedulable.

Reviewer: Bolded the list intros, cut "it", and made the criteria for moving pods to the unschedulable pool a sub-list for scannability.

## Scheduling framework and plugins

The Kubernetes scheduler is implemented following the Kubernetes
[scheduling framework](/docs/concepts/scheduling-eviction/scheduling-framework/).

> And, each scheduling requirements are implemented as a plugin.

Reviewer 2: Consider rephrasing to "all scheduling features are implemented as plugins"

and lastly decides in which Node to run the Pod, or concludes that the Pod cannot go to anywhere for now.

If the scheduling is successful, the second phase, called the _binding cycle_, binds the Pod with
the Node by communicating the decision to the API server.
But, if it turns out that the Pod cannot go to anywhere during the scheduling cycle,
the binding cycle isn't executed; instead the Pod is moved back to the scheduling queue.
Although there are some exceptions, unscheduled Pods enter the _unschedulable pod pool_.

Pods in Unschedulable Pod Pool are moved to ActiveQ/BackoffQ 
only when Scheduling Queue identifies changes in the cluster that might be schedulable if we retry the scheduling.

That is a crucial step because scheduling cycle is performed for Pods one by one -
if we didn't have Unschedulable Pod Pool and kept retrying the scheduling of any Pods, 
multiple scheduling cycles would be wasted for Pods that have no chance to be scheduled.

Then, how do they decide when to move a Pod back into the ActiveQ? How do they notice that Pods might be schedulable now?

> Here QueueingHints come into play.

Reviewer 2: You could say that so far any event was triggering reattempts, but the queuing hints brought huge improvement here. Otherwise, the reader needs to read further details to figure it out.

> The first phase, called the _scheduling cycle_, takes Pods from activeQ **one by one**, runs all plugins' logic,
> and lastly decides in which Node to run the Pod, or concludes that the Pod cannot go to anywhere for now.
> 
> If the scheduling is successful, the second phase, called the _binding cycle_, binds the Pod with
> the Node by communicating the decision to the API server.
> But, if it turns out that the Pod cannot go to anywhere during the scheduling cycle,
> the binding cycle isn't executed; instead the Pod is moved back to the scheduling queue.
> Although there are some exceptions, unscheduled Pods enter the _unschedulable pod pool_.
> 
> Pods in Unschedulable Pod Pool are moved to ActiveQ/BackoffQ 
> only when Scheduling Queue identifies changes in the cluster that might be schedulable if we retry the scheduling.
> 
> That is a crucial step because scheduling cycle is performed for Pods one by one -
> if we didn't have Unschedulable Pod Pool and kept retrying the scheduling of any Pods, 
> multiple scheduling cycles would be wasted for Pods that have no chance to be scheduled.
> 
> Then, how do they decide when to move a Pod back into the ActiveQ? How do they notice that Pods might be schedulable now?
> Here QueueingHints come into play.

Reviewer: ```suggestion

The scheduler processes pending Pods in phases called _cycles_ as follows:

1. **Scheduling cycle**: the scheduler takes pending Pods from the activeQ component of the scheduling
    queue  _one by one_. For each Pod,  the scheduler runs the logic from every scheduling plugin. The
    scheduler then decides on a node for the Pod or decides that the Pod can't be scheduled at that time.
    
    If the scheduler decides that a Pod can't be scheduled, that Pod enters the Unschedulable Pod Pool
    component of the scheduling queue. However, if the scheduler decides to place the Pod on a node, 
    the next cycle executes for that Pod.
    
1. **Binding cycle**: the scheduler communicates the node placement decision to the Kubernetes API
    server. The Pod is then bound to the selected node. 
    
Aside from some exceptions, most unscheduled Pods enter the unschedulable pod pool after each scheduling
cycle. The Unschedulable Pod Pool component is crucial because of how the scheduling cycle processes Pods one by one. If the scheduler had to constantly retry placing unschedulable Pods instead of offloading those
Pods to the Unschedulable Pod Pool, multiple scheduling cycles would be wasted on those Pods.
```

> ## QueueingHint
> 
> QueueingHint is callback function per plugin to notice an object addition/update/deletion in the cluster (we call them cluster events)
> that may make Pods schedulable.
> 
> Let's say the Pod `pod-a` has a required Pod affinity, and got rejected in scheduling cycle by the `PodAffinity` plugin
> because no Node has any Pod matching the Pod affinity specification for `pod-a`.
> 
> ![pod-a got rejected by PodAffinity](./queueinghint1.svg)
> 
> When an unscheduled Pod is put into the unschedulable pod pool, the scheduling queue
> records which plugins caused the scheduling failure of the Pod.
> In this example, scheduling queue notes that `pod-a` was rejected by `PodAffinity`.
> 
> `pod-a` will never be schedulable until the PodAffinity failure is resolved somehow.
> The scheduling queue uses the queueing hints from plugins that rejected the Pod, which is `PodAffinity` in the example.
> 
> A QueueingHint subscribes to a particular kind of cluster event and make a decision whether an incoming event could make the Pod schedulable.
> Thinking about when PodAffinity failure could be resolved,
> one possible scenario is that an existing Pod gets a new label which matches with `pod-a`'s PodAffinity.
> 
> The `PodAffinity` plugin's `QueueingHint` callback checks on all Pod updates happening in the cluster, 
> and when it catches such update, the scheduling queue moves `pod-a` to either ActiveQ or BackoffQ.
> 
> ![pod-a is moved by PodAffinity QueueingHint](./queueinghint2.svg)
> 
> We actually already had a similar functionality (called `preCheck`) inside the scheduling queue,
> which filters out cluster events based on Kubernetes core scheduling constraints - 
> for example, filtering out node related events when nodes aren't ready.
> 
> But, it's not ideal because this hard-coded `preCheck` refers to in-tree plugins logic,
> and it causes issues for custom plugins (for example: [#110175](https://github.com/kubernetes/kubernetes/issues/110175)).
> 

Reviewer: ```suggestion
## Improvements to retrying Pod scheduling with QueuingHint

Unschedulable Pods only move back into the ActiveQ or BackoffQ components of the scheduler
queue if changes in the cluster might allow the scheduler to place those Pods on nodes. Prior to v1.32, 
the scheduling queue identified these changes in the cluster by using a feature called `preCheck`. This
built-in functionality filters out cluster events based on Kubernetes core scheduling constraints. For example,
`preCheck` could filter out node-related events when the node status is `NotReady`. 

However, `preCheck` relies on the logic of in-tree plugins and caused some issues with custom plugins,
like in issue [#110175](https://github.com/kubernetes/kubernetes/issues/110175). Additionally, the broad
scope of event filtering could lead to scheduling retries for no reason. For example, the
scheduling queue might retry scheduling after a node update event in case the update created more
resources for the previously unschedulable Pod. However, this would result in a retry attempt for every
node update, even if those updates didn't actually change resource capacity on the nodes. 

QueueingHint is a per-plugin callback function to notice an object creation, update, or deletion in the cluster,
called _cluster events_, that might make unschedulable Pods schedulable. A QueueingHint subscribes to a 
particular kind of cluster event and make a decision about whether an incoming event could make the Pod
schedulable. The scheduling queue uses the queueing hints from plugins that rejected the Pod to decide
whether to retry scheduling for that Pod. 

For example, consider a Pod named `pod-a` that has a required Pod affinity. `pod-a` was rejected in
the scheduling cycle by the `PodAffinity` plugin because no node had an existing Pod that matched
the Pod affinity specification for `pod-a`.

![pod-a got rejected by PodAffinity](./queueinghint1.svg)

`pod-a` moves into the Unschedulable Pod Pool. The scheduling queue records which plugin caused
the scheduling failure for the Pod. For `pod-a`, the scheduling queue records that the `PodAffinity`
plugin rejected the Pod.

`pod-a` will never be schedulable until the PodAffinity failure is resolved. The `PodAffinity` plugin's
`QueuingHint` callback function checks every Pod update that occurs in the cluster. If, for example,
a Pod gets a label update that matches the Pod affinity requirement of `pod-a`, the `PodAffinity`
plugin's `QueuingHint` prompts the scheduling queue to move `pod-a` back into the ActiveQ or
the BackoffQ component.

![pod-a is moved by PodAffinity QueueingHint](./queueinghint2.svg)
```

Author: Thanks, but some explanations are not quite correct and I had to modify.

## QueueingHint's history and what's new in v1.32

Within SIG Scheduling, we have been working on the development of QueueingHint since
Kubernetes v1.28.

> QueueingHint is not something user-facing, but we implemented a feature gate (`SchedulerQueueingHints`) as a safety net,
> ...which actually saved our life soon.
> 
> In v1.28, we implemented QueueingHints with a few in-tree plugins experimentally,
> and made the feature gate enabled by default.
> 
> But, users reported the memory leak issue, and consequently we disabled the feature gate in the patch release of v1.28.
> 
> In v1.28 - v1.31, we kept working on the QueueingHint implementation within the rest of in-tree plugins,
> and having bug fixes.

Reviewer: ```suggestion
While QueuingHint isn't user-facing, we implemented the `SchedulerQueueingHints` feature gate as a
safety measure when we originally added this feature. In v1.28, we implemented QueueingHints with a
few in-tree plugins experimentally, and made the feature gate enabled by default.

However, users reported a memory leak issue, and consequently we disabled the feature gate in a
patch release of v1.28.  From v1.28 until v1.31, we kept working on the QueueingHint implementation
within the rest of the in-tree plugins and fixing bugs.
```

> And, at v1.32, we will make this feature enabled by default again;
> we finished implementing QueueingHints with all plugins,
> and also identified the cause of the memory leak, finally!

Reviewer: ```suggestion
In v1.32, we will make this feature enabled by default again. We finished implementing QueueingHints
in all plugins and also identified the cause of the memory leak!
```

    queue  _one by one_. For each Pod, the scheduler runs the filtering/scoring logic from every scheduling plugin. The
    scheduler then decides on the best node for the Pod, or decides that the Pod can't be scheduled at that time.
    
    If the scheduler decides that a Pod can't be scheduled, that Pod enters the Unschedulable Pod Pool
    component of the scheduling queue. However, if the scheduler decides to place the Pod on a node, 
    the next cycle executes for that Pod.
    
1. **Binding cycle**: the scheduler communicates the node placement decision to the Kubernetes API
    server. The Pod is then bound to the selected node. 
    
Aside from some exceptions, most unscheduled Pods enter the unschedulable pod pool after each scheduling
cycle. The Unschedulable Pod Pool component is crucial because of how the scheduling cycle processes Pods one by one. If the scheduler had to constantly retry placing unschedulable Pods instead of offloading those
Pods to the Unschedulable Pod Pool, multiple scheduling cycles would be wasted on those Pods.

## Improvements to retrying Pod scheduling with QueuingHint

> Unschedulable Pods only move back into the ActiveQ or BackoffQ components of the scheduler
> queue if changes in the cluster might allow the scheduler to place those Pods on nodes. 

Reviewer 3: ```suggestion
Unschedulable Pods only move back into the ActiveQ or BackoffQ components of the scheduling
queue if changes in the cluster might allow the scheduler to place those Pods on nodes. 
```

- `preCheck` relied on the logic of in-tree plugins and caused some issues for custom plugins,
like in issue [#110175](https://github.com/kubernetes/kubernetes/issues/110175). 

Here QueueingHints come into play; 
a QueueingHint subscribes to a particular kind of cluster event, and make a decision about whether each incoming event could make the Pod schedulable.

For example, consider a Pod named `pod-a` that has a required Pod affinity. `pod-a` was rejected in
the scheduling cycle by the `InterPodAffinity` plugin because no node had an existing Pod that matched
the Pod affinity specification for `pod-a`.

![pod-a got rejected by InterPodAffinity](./queueinghint1.svg)

`pod-a` moves into the Unschedulable Pod Pool. The scheduling queue records which plugin caused
the scheduling failure for the Pod. For `pod-a`, the scheduling queue records that the `InterPodAffinity`
plugin rejected the Pod.

`pod-a` will never be schedulable until the InterPodAffinity failure is resolved. The `InterPodAffinity` plugin's
`QueuingHint` callback function checks every Pod update that occurs in the cluster. If, for example,

> a Pod gets a label update that matches the Pod affinity requirement of `pod-a`, the `InterPodAffinity`

Reviewer 3: I won't say that _"`InterPodAffinity` plugin's
`QueuingHint` callback function checks every Pod update"_ as it only checks the UpdatePodLabel event, to be precise.

Author: Modified the explanation around here a bit.

Unschedulable Pods only move back into the ActiveQ or BackoffQ components of the scheduling
queue if changes in the cluster might allow the scheduler to place those Pods on nodes. 

Prior to v1.32, each plugin registered which cluster changes could solve their failures, an object creation, update, or deletion in the cluster (called _cluster events_),
with `EnqueueExtensions` (`EventsToRegister`),
and the scheduling queue retries a pod with an event that is registered by a plugin that rejected the pod in a previous scheduling cycle.

Additionally, we had an internal feature called `preCheck`, which helped further filtering of events for efficiency, based on Kubernetes core scheduling constraints;
For example, `preCheck` could filter out node-related events when the node status is `NotReady`. 

However, we had two issues for those approaches:
- Requeueing with events was too broad, could lead to scheduling retries for no reason.
  - For example, a new scheduled Pod _might_ solve the `InterPodAffinity`'s failure, but not all of them do, 
for example, if a new Pod is created, but without a label matching `InterPodAffinity` of the unschedulable pod, the pod wouldn't be schedulable.

> - `preCheck` relied on the logic of in-tree plugins and caused some issues for custom plugins,

Author: `to` sounds more natural for me. (not sure either)

like in issue [#110175](https://github.com/kubernetes/kubernetes/issues/110175). 

Here QueueingHints come into play; 
a QueueingHint subscribes to a particular kind of cluster event, and make a decision about whether each incoming event could make the Pod schedulable.

For example, consider a Pod named `pod-a` that has a required Pod affinity. `pod-a` was rejected in
the scheduling cycle by the `InterPodAffinity` plugin because no node had an existing Pod that matched
the Pod affinity specification for `pod-a`.

> ![pod-a got rejected by InterPodAffinity](./queueinghint1.svg)

Reviewer 4: _Improvement:_ To display the image incorporate a '[figure](https://gohugo.io/content-management/shortcodes/#figure)' shortcode with a caption and maybe also include an alt attribute for better accessibility. Take a look at how the shortcode is utilized in an [old blog post (here)](https://github.com/kubernetes/website/blob/main/content/en/blog/_posts/2024-04-05-diy-create-your-own-cloud-with-kubernetes-part-1/index.md?plain=1#L59).

`pod-a` moves into the Unschedulable Pod Pool. The scheduling queue records which plugin caused
the scheduling failure for the Pod. For `pod-a`, the scheduling queue records that the `InterPodAffinity`
plugin rejected the Pod.

`pod-a` will never be schedulable until the InterPodAffinity failure is resolved. 
There're some scenarios that the failure could be resolved, one example is an existing running pod gets a label update and becomes matching a Pod affinity.
For this scenario, the `InterPodAffinity` plugin's `QueuingHint` callback function checks every Pod label update that occurs in the cluster. 
Then, if a Pod gets a label update that matches the Pod affinity requirement of `pod-a`, the `InterPodAffinity`,
plugin's `QueuingHint` prompts the scheduling queue to move `pod-a` back into the ActiveQ or
the BackoffQ component.

> ![pod-a is moved by InterPodAffinity QueueingHint](./queueinghint2.svg)

Reviewer 4: Again here would be nice to use a '[figure](https://gohugo.io/content-management/shortcodes/#figure)' shortcode.

_Refer an [old blog post (here)](https://github.com/kubernetes/website/blob/main/content/en/blog/_posts/2024-04-05-diy-create-your-own-cloud-with-kubernetes-part-1/index.md?plain=1#L59) which used the `figure` shortcode._

## QueueingHint's history and what's new in v1.32

> Within SIG Scheduling, we have been working on the development of QueueingHint since

Reviewer 5: ```suggestion
At SIG Scheduling, we have been working on the development of QueueingHint since
```

Kubernetes v1.28.

While QueuingHint isn't user-facing, we implemented the `SchedulerQueueingHints` feature gate as a
safety measure when we originally added this feature. In v1.28, we implemented QueueingHints with a
few in-tree plugins experimentally, and made the feature gate enabled by default.

> However, users reported a memory leak issue, and consequently we disabled the feature gate in a

Reviewer 5: ```suggestion
However, users reported a memory leak, and consequently we disabled the feature gate in a
```

patch release of v1.28.  From v1.28 until v1.31, we kept working on the QueueingHint implementation
within the rest of the in-tree plugins and fixing bugs.

> In v1.32, we will make this feature enabled by default again. We finished implementing QueueingHints

Reviewer 5: ```suggestion
In v1.32, we made this feature enabled by default again. We finished implementing QueueingHints
```

in all plugins and also identified the cause of the memory leak!

> 

Reviewer 5: Maybe add something like this:

We thank all the contributors who participated in the development of this feature and those who reported and investigated the earlier issues.

---
layout: blog
title: "Kubernetes v1.32: QueueingHint Brings a New Possibility to Optimize Pod Scheduling"
date: 2024-12-12
slug: scheduler-queueinghint

> draft: true

Reviewer 6: I was trying to find the preview to review and could not find it and this line should be removed for this to be published
```suggestion
```

Reviewer 7: I'll DM you on Slack about this parameter, it's a new way of scheudling blogs we orchestrated with @kubernetes/sig-docs-blog-owners

Author: >
  [Kensei Nakada](https://github.com/sanposhiho) (Tetrate.io)
---

The Kubernetes [scheduler](/docs/concepts/scheduling-eviction/kube-scheduler/) is the core
component that selects the nodes on which new Pods run. The scheduler processes
these new Pods **one by one**. Therefore, the larger your clusters, the more important
the throughput of the scheduler becomes.

> Over the years, the Kubernetes project (and SIG Scheduling in particular) has improved the throughput

Reviewer 6: non-blocking nit / style suggestion
```suggestion
Over the years, Kubernetes SIG Scheduling improved the throughput
```

Author: opened: https://github.com/kubernetes/website/pull/48956
